gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java

```diff
@@ -414,30 +415,49 @@ public void run() {
       }
     }
 
+    /**
+     * Cancels the dag and sends a cancellation tracking event.
+     * @param dagToCancel dag node to cancel
+     * @throws ExecutionException executionException
+     * @throws InterruptedException interruptedException
+     */
     private void cancelDag(String dagToCancel) throws ExecutionException, InterruptedException {
       log.info("Cancel flow with DagId {}", dagToCancel);
       if (this.dagToJobs.containsKey(dagToCancel)) {
         List<DagNode<JobExecutionPlan>> dagNodesToCancel = this.dagToJobs.get(dagToCancel);
         log.info("Found {} DagNodes to cancel.", dagNodesToCancel.size());
         for (DagNode<JobExecutionPlan> dagNodeToCancel : dagNodesToCancel) {
-          Properties props = new Properties();
-          if (dagNodeToCancel.getValue().getJobFuture().isPresent()) {
-            Future future = dagNodeToCancel.getValue().getJobFuture().get();
-            if (future instanceof CompletableFuture &&
-              future.get() instanceof AzkabanExecuteFlowStatus.ExecuteId) {
-              CompletableFuture<AzkabanExecuteFlowStatus.ExecuteId> completableFuture = (CompletableFuture) future;
-              String azkabanExecId = completableFuture.get().getExecId();
-              props.put(ConfigurationKeys.AZKABAN_EXEC_ID, azkabanExecId);
-              log.info("Cancel job with azkaban exec id {}.", azkabanExecId);
-            }
-          }
-          DagManagerUtils.getSpecProducer(dagNodeToCancel).deleteSpec(null, props);
+          cancelDag(dagNodeToCancel);
         }
       } else {
         log.warn("Did not find Dag with id {}, it might be already cancelled.", dagToCancel);
       }
     }
 
+    private void cancelDag(DagNode<JobExecutionPlan> dagNodeToCancel) throws ExecutionException, InterruptedException {
+      Properties props = new Properties();
+      if (dagNodeToCancel.getValue().getJobFuture().isPresent()) {
+        Future future = dagNodeToCancel.getValue().getJobFuture().get();
+        if (future instanceof CompletableFuture &&
+            future.get() instanceof AzkabanExecuteFlowStatus.ExecuteId) {
```

Comment:
Aren't we leaking Azkaban related details into the DagManager? Why can't this be abstracted away?

Reply:
So, should I change the signature from SpecProducer.deleteSpec(URI, Properties) to SpecProducer.deleteSpec(Future) ?

Reply:
Why can't the future be serialized into a string and passed via Properties, if all you care about is passing execution id.